server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/MergedPointRangeQuery.java

```diff
@@ -0,0 +1,207 @@
+                    return null;
+                }
+                return scorerSupplier.get(Long.MAX_VALUE);
+            }
+
+            @Override
+            public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOException {
+                /*
+                 * If we're sure docs only have a single value for the field
+                 * we can pick the most compact bounds. If there are multiple values
+                 * for the field we have to run the boolean query.
+                 */
+                PointValues points = context.reader().getPointValues(field);
+                if (points == null) {
+                    return null;
+                }
+                if (points.size() == points.getDocCount()) {
+                    // Each doc that has points has exactly one point.
+                    return singleValuedSegmentWeight().scorerSupplier(context);
+                }
+                return multiValuedSegmentWeight().scorerSupplier(context);
+            }
+
+            @Override
+            public BulkScorer bulkScorer(LeafReaderContext context) throws IOException {
+                PointValues points = context.reader().getPointValues(field);
+                if (points == null) {
+                    return null;
+                }
+                if (points.size() == points.getDocCount()) {
+                    // Each doc that has points has exactly one point.
+                    return singleValuedSegmentWeight().bulkScorer(context);
+                }
+                return multiValuedSegmentWeight().bulkScorer(context);
+            }
+
+            private Weight singleValuedSegmentWeight() throws IOException {
+                if (singleValuedSegmentWeight == null) {
+                    singleValuedSegmentWeight = delegateForSingleValuedSegments.createWeight(searcher, scoreMode, boost);
+                }
+                return singleValuedSegmentWeight;
+            }
+
+            private Weight multiValuedSegmentWeight() throws IOException {
+                if (multiValuedSegmentWeight == null) {
+                    multiValuedSegmentWeight = delegateForMultiValuedSegments.createWeight(searcher, scoreMode, boost);
+                }
+                return multiValuedSegmentWeight;
+            }
+        };
+    }
+
+    /**
+     * The query used when we have single valued segments.
+     */
+    Query delegateForSingleValuedSegments() {
+        return delegateForSingleValuedSegments;
+    }
+
+    @Override
+    public String toString(String field) {
+        return "MergedPointRange[" + delegateForMultiValuedSegments.toString(field) + "]";
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        MergedPointRangeQuery other = (MergedPointRangeQuery) obj;
+        return delegateForMultiValuedSegments.equals(other.delegateForMultiValuedSegments);
```

Comment:
I think this is incomplete, we should check delegateForSingleValuedSegments as well?

Reply:
They depend on one another so I don't think we *technically* have to, but I agree, we may as well.